geomstats/vectorization.py

```diff
@@ -112,59 +28,171 @@ def decorator(point_types):
+        List of inputs' input_types, including for optional inputs.
+        The `input_type`s of optional inputs will not be read
         by the decorator if the corresponding input is not given.
     """
-    if not isinstance(point_types, list):
-        point_types = list(point_types)
+    if not isinstance(input_types, list):
+        input_types = list(input_types)
 
     def aux_decorator(function):
         def wrapper(*args, **kwargs):
-            args_point_types = point_types[:len(args)]
-            kwargs_point_types = point_types[len(args):]
+            args_types, kwargs_types, opt_kwargs_types, is_scal = get_types(
+                input_types, args, kwargs)
+
+            args_types, kwargs_types, kwargs = adapt_types(
+                args_types, kwargs_types, opt_kwargs_types, args, kwargs)
+            args_kwargs_types = args_types + kwargs_types
 
-            in_shapes = initial_shapes(args_point_types, args)
-            kw_in_shapes = initial_shapes(kwargs_point_types, kwargs.values())
-            in_shapes.extend(kw_in_shapes)
+            args_shapes = get_initial_shapes(args_types, args)
+            kwargs_shapes = get_initial_shapes(kwargs_types, kwargs.values())
+            initial_shapes = args_shapes + kwargs_shapes
 
-            vect_args = vectorize_args(args_point_types, args)
-            vect_kwargs = vectorize_kwargs(kwargs_point_types, kwargs)
+            vect_args = vectorize_args(args_types, args)
+            vect_kwargs = vectorize_kwargs(kwargs_types, kwargs)
 
             result = function(*vect_args, **vect_kwargs)
 
-            if squeeze_output_dim_1(result, in_shapes, point_types):
-                if result.shape[1] == 1:
-                    result = gs.squeeze(result, axis=1)
+            result = adapt_result(
+                result, initial_shapes, args_kwargs_types, is_scal)
 
-            if squeeze_output_dim_0(in_shapes, point_types):
-                if result.shape[0] == 1:
-                    result = gs.squeeze(result, axis=0)
             return result
         return wrapper
     return aux_decorator
 
 
-def initial_shapes(point_types, args):
+def get_types(input_types, args, kwargs):
+    """Extract the types of args, kwargs, optional kwargs and output.
+
+    Parameters
+    ----------
+    input_types : list
+        List of inputs' input_types, including for optional inputs.
+    args : tuple
+        Args of a function.
+    kwargs : dict
+        Kwargs of a function.
+
+    Returns
+    -------
+    args_types : list
+        Types of args.
+    kwargs_types : list
+        Types of kwargs.
+    opt_kwargs_types : list
+        Types of optional kwargs.
+    is_scal : bool
+        Boolean determining if the output is a scalar.
+    """
+    len_args = len(args)
+    len_kwargs = len(kwargs)
+    len_total = len_args + len_kwargs
+
+    args_types = input_types[:len_args]
+    kwargs_types = input_types[len_args:len_total]
+
+    opt_kwargs_types = []
```

Comment:
I wonder if you could do this with the np.split method?

Reply:
Yes, but it wouldn't actually save a line, because I first need to convert `input_types` to an array, and then use several splits because of the if/else.